File: APM/Java/Interpreter/src/ro/sandorrobertk94/domain/MyDictionary.java
package ro.sandorrobertk94.domain;

import java.util.Arrays;
import java.util.Objects;

/**
 * Created by robert on 10/25/15.
 */
public class MyDictionary implements MyIDictionary {
    private static final int CAPACITY = 50;
    private int size;
    private String[] keys;
    private Object[] values;

    public MyDictionary() {
        size = 0;
        keys = new String[CAPACITY];
        values = new Object[CAPACITY];
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public void add(String key, Object value) {
        int i;
        for (i = 0; i < size; i++) {
            if (Objects.equals(key, keys[i])) {
                break;
            }
        }

        // if we found the key
        if (i < size) {
            values[i] = value;
        } else {
            keys[size] = key;
            values[size] = value;
            size++;
        }
    }

    @Override
    public Object lookUp(String key) throws Exception {
        int i;
        for (i = 0; i < size; i++) {
            if (Objects.equals(key, keys[i])) {
                break;
            }
        }

        // if we found the key
        if (i < size) {
            return values[i];
        } else {
            throw new Exception("Key not found");
        }
    }

    @Override
    public String toString() {
        String out = "{ size = " + size + "\n";
        for (int i = 0; i < size; i++) {
            out += keys[i] + " => " + values[i].toString() + "\n";
        }
        out += "};";
        return out;
    }
}

File: APM/CSharp/Interpreter/Interpreter/domain/MyDictionary.cs
using System;

namespace Interpreter.domain
{
    public class MyDictionary : MyIDictionary
    {
        private const int Capacity = 50;
        private readonly string[] _keys;
        private int _size;
        private readonly object[] _values;

        public MyDictionary()
        {
            _size = 0;
            _keys = new string[Capacity];
            _values = new object[Capacity];
        }


        public bool IsEmpty()
        {
            return _size == 0;
        }

        public void Add(string key, object value)
        {
            int i;
            for (i = 0; i < _size; i++)
            {
                if (key == _keys[i])
                {
                    break;
                }
            }

            if (i < _size)
            {
                _values[i] = value;
            }
            else
            {
                _keys[_size] = key;
                _values[_size] = value;
                _size++;
            }
        }

        public object LookUp(string key)
        {
            int i;
            for (i = 0; i < _size; i++)
            {
                if (key == _keys[i])
                {
                    break;
                }
            }

            if (i < _size)
            {
                return _values[i];
            }
            throw new Exception("Key not found!");
        }

        public override string ToString()
        {
            var o = "{ size = " + _size + "\n";
            for (var i = 0; i < _size; i++)
            {
                o += _keys[i] + " => " + _values[i] + "\n";
            }
            o += "};";
            return o;
        }
    }
}